### DISCUSSION TITLE
How to use cosine similarity search with milvus?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#20531] [Q&A and General discussion]
How to use cosine similarity search with milvus?

How to normalize already existing vectors?

**Comment by yhmo** [comment]
As we know:
![Screenshot from 2022-11-14 14-44-22](https://user-images.githubusercontent.com/2282099/201594591-74945e93-6925-423a-90a4-c37ee801cced.png)

if ||A|| = 1 and ||B|| = 1, then cos(θ) = AB. 
AB is the inner product. So, when A and B are both normalized, their length is 1, the similarity = cos(θ) = inner product of A and B

For float vectors, Milvus supports two kinds of metric types: "L2" and "IP"
To search with inner product:
```
search_param = {
        "data": search_vectors,
        "anns_field": vector_field,
        "param": {"metric_type": "IP", "params": {"nprobe": _NPROBE}},
        "limit": _TOPK}
    results = collection.search(**search_param)
```

**But milvus doesn't normalize the vectors, users need to normalize vectors before inserting them into milvus.**
To normalize vector by python, please refer this article: https://www.delftstack.com/howto/numpy/normalize-vector-in-python/

**Reply by martinenkoEduard** [reply]
But if you try

import numpy as np

v = np.random.rand(10)

normalized_v = v / np.sqrt(np.sum(v**2))
print(normalized_v)

print(np.sum(normalized_v))

sum is not 1

**Reply by yhmo** [reply]
> But if you try
> 
> import numpy as np
> 
> v = np.random.rand(10)
> 
> normalized_v = v / np.sqrt(np.sum(v**2)) print(normalized_v)
> 
> print(np.sum(normalized_v))
> 
> sum is not 1

After normalize, the length of vector is 1.0. Assume the normalized vector is [x1, x2, x3, x4, x5 ....], its length is **sqrt(x1*x1 + x2*x2 + x3*x3 + x4*x4 + ...)**
Change the last line to:  `print(np.sqrt((normalized_v*normalized_v).sum()))`, you will get 1.0
